### ISSUE TITLE
upgrade playbook restarts docker on all nodes that changed docker config simultanously

### ISSUE THREAD

**Issue by ibotty** [openshift/openshift-ansible#2171]
That happens when docker configuration changes in the docker role. This role transitively gets pulled in by openshift facts (afaict).

**Comment by dgoodwin** [comment]
I think it's probably openshift_version which is responsible for determining the precise version we're installing or upgrading to. For containerized installs this requires docker to lookup the version from the containers on the registry in some cases.

@ibotty could you elaborate on how you hit this? I'm curious if you know exactly what was changed in Docker config, whether it was a change in your inventory or if it was something in the playbooks. What version and deployment type were you upgrading? Would you be able to share your inventory variables as well?

@detiber bit of a jam on this one, how do you feel about a flag that skips all docker config changes (if it's already installed and running) so openshift_version could still require docker to be running and check the appropriate versions.

**Comment by ibotty** [comment]
It was modifying /etc/sysconfig/docker's "OPTIONS=". (interestingly I now have --insecure-registry in OPTIONS as well as in INSECURE_REGISTRY).

**Comment by ibotty** [comment]
@dgoodwin It was not the inventory afaict. I am not the only one administrating that test cluster, so I don't know for certain though.

**Comment by dgoodwin** [comment]
@detiber alternative proposal, what if we remove the automatic restart on docker config changes, it's very dangerous to have floating around with the way things are today. If you do anything that modified docker config, and re-run config.yml, all your nodes and masters go down at once in parallel.

I'm guessing we don't want to make config fully serialized.

Upgrade is already serialized for nodes, but per this issue the docker role does get used elsewhere prior to that, so if there's a pending docker config change incoming everything goes down at once.

FWIW, in my testing docker does not restart in upgrading to 3.2 so I think something may have changed in this environment or inventory that triggered the config change. Normally, that should not happen.

Regardless you can't safely roll out a docker config change with config.yml re-runs today, it will execute in parallel.

**Comment by ibotty** [comment]
I had a parallel docker restart at another client when upgrading from 1.1.6 to 1.2 also with docker config changes.

Given that ansible/ansible#12170 is not going to land soonish, what do you think about the following approach for docker handlers?

```yaml
- hosts: all
  handlers:
  - name: restart docker
    set_fact: restart_docker=yes
  tasks:
  ...

- hosts: hosts with restart_docker=yes
  serial 1:
  tasks:
  - isolate
  - restart docker
  - ...
```

Of course that requires that every playbook using the docker role (or any restart docker) uses this part. It should be made into a playbook to be included I'd say.

**Comment by dgoodwin** [comment]
@ibotty interesting idea, the issue is this role is used by both config playbook (which does not yet know how to do much serially, and your solution would be prefect for), and by upgrade playbooks which as of recently are carefully evacuating nodes and then handling docker and node upgrades together, all of which is done serially. It will be a bit complicated to control the behavior so we can leveage your idea above in some situations, but not in others, given how easy it is to accidentally include the docker role via some layers dependency hierarchy. Will give it some thought because I like this idea for config playbook.

**Comment by ibotty** [comment]
I have used that approach a few times and you mostly lose modularity via roles, because you will need to know which roles need this special handling.

I slowly converged to the convention that a role with such a serialized handler has a "handler_tasks" subdirectory which includes the handlers (one file for each) and a subdirectory "handler_plays" with one file per handler which evaluates "the groups that need handling" and one that includes the tasks in the handler_tasks dir.

**Comment by detiber** [comment]
@ibotty so, we did add an additional '--insecure-registry' to OPTIONS to cover the internally deployed registry without complicating the INSECURE_REGISTRY setting defined by the user.

This is definitely a tricky problem to solve, since removing the handler from the role would require special handling to make sure that the handler is applied appropriately.

@dgoodwin since the issue mainly is going to crop up on upgrades, I'm wondering if we just need to apply the openshift_docker role serially prior to doing other steps to ensure that we don't restart docker in parallel at a later step?

**Comment by dgoodwin** [comment]
@detiber I don't think that works out either due to my new openshift_version dep on docker, that is part of pre-upgrade which would imply pre-upgrade has to go serial, which doesn't work at all for it's purpose of making sure everything's ok before proceeding. Even if we solve that we'd be introducing a double node evac for every node.

**Comment by dgoodwin** [comment]
FWIW, I already introduced a docker_protect_installed_version variable, which defaults to False, but is explicitly set to True only during an upgrade pre.yml. (where we just need information and want to be very careful not to mess with the system yet)

That is basically the exact situation we're in here, I think if I just guard the restart with that variable, we would effectively disable the restart during upgrade pre.yml, and it would be restarted later when we actually upgrade docker.

**Comment by detiber** [comment]
@dgoodwin I don't think I'm a fan of double-purposing that variable...  if we need a variable to prevent a restart, I would almost prefer that it be a separate variable explicitly for that purpose.

**Comment by dgoodwin** [comment]
A new variable would be a good quick solution. I do have the nagging feeling the intent is the same, use the role to ensure docker is installed and running, but don't interrupt it or upgrade it. (Could also rename it something to better reflect that, or introduce a new role)

I'll go the new var route though as that is very simple, and should correct this for now. We can also stop worrying about it with 1.12.

**Comment by dgoodwin** [comment]
A fix for this was merged in #2441 , docker role should now be skipped entirely until we're ready to actually allow for the possibility of restarting docker. (i.e. node has been evacuated)
